File: src/main/java/me/partlysanestudios/partlysaneskies/render/gui/components/PSSButton.java
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.render.gui.components;

import cc.polyfrost.oneconfig.config.core.OneColor;
import gg.essential.elementa.UIComponent;
import gg.essential.elementa.components.UIBlock;
import gg.essential.elementa.components.UIImage;
import gg.essential.elementa.components.UIWrappedText;
import gg.essential.elementa.constraints.CenterConstraint;
import gg.essential.elementa.constraints.PixelConstraint;
import gg.essential.elementa.constraints.XConstraint;
import gg.essential.elementa.constraints.YConstraint;
import gg.essential.elementa.events.UIClickEvent;
import me.partlysanestudios.partlysaneskies.features.themes.ThemeManager;

import java.awt.*;
import java.util.function.Consumer;

public class PSSButton {
    UIBlock backgroundBlock;
    UIImage buttonTexture;
    UIWrappedText textComponent;

    float width;
    float height;
    XConstraint xConstraint;
    YConstraint yConstraint;
    OneColor color;
    String text;

    public PSSButton() {
        this.text = "";

        backgroundBlock = (UIBlock) new UIBlock()
            .setColor(new Color(0, 0, 0, 0));
        
        buttonTexture = (UIImage) ThemeManager.getCurrentButtonUIImage()
            .setChildOf(backgroundBlock);

        textComponent = (UIWrappedText) new UIWrappedText(text, false, new Color(0, 0, 0, 0), true)
            .setColor(new Color(255, 255, 255, 255))
            .setChildOf(buttonTexture);
        
        this.color = new OneColor(0, 0, 0, 0);
    }

    public PSSButton(Color color){
        this(new OneColor(color));
    }

    public PSSButton(OneColor color) {
        this.text = "";

        this.backgroundBlock = (UIBlock) new UIBlock()
            .setColor(color.toJavaColor());
    
        this.buttonTexture = (UIImage) ThemeManager.getCurrentButtonUIImage(color)
            .setChildOf(backgroundBlock);

        this.textComponent = (UIWrappedText) new UIWrappedText(text, false, new Color(0, 0, 0, 0), true)
            .setColor(new Color(255, 255, 255, 255))
            .setChildOf(buttonTexture);

        this.color = color;
    }


    public PSSButton setHeight(float height) {
        backgroundBlock.setHeight(new PixelConstraint(height));
        buttonTexture.setHeight(new PixelConstraint(height));

        this.height = height;

        return this;
    }

    public PSSButton setWidth(float width) {
        backgroundBlock.setWidth(new PixelConstraint(width));
        buttonTexture.setWidth(new PixelConstraint(width));
        textComponent.setWidth(new PixelConstraint(width));

        this.width = width;

        return this;
    }

    public PSSButton setX(XConstraint xPos) {
        backgroundBlock.setX(xPos);
        buttonTexture.setX(new CenterConstraint());
        textComponent.setX(new CenterConstraint());

        this.xConstraint = xPos;

        return this;
    }

    public PSSButton setY(YConstraint yPos) {
        backgroundBlock.setY(yPos);
        buttonTexture.setY(new CenterConstraint());
        textComponent.setY(new CenterConstraint());
        
        this.yConstraint = yPos;

        return this;
    }

    public PSSButton setChildOf(UIComponent parent) {
        backgroundBlock.setChildOf(parent);

        return this;
    }

    public PSSButton setColor(Color color) {
        return setColor(new OneColor(color));
    }
    public PSSButton setColor(OneColor color) {
        backgroundBlock.removeChild(buttonTexture);

        buttonTexture = (UIImage) ThemeManager.getCurrentButtonUIImage(color)
            .setWidth(new PixelConstraint(this.width))
            .setHeight(new PixelConstraint(this.height))
            .setX(new CenterConstraint())
            .setY(new CenterConstraint())
            .setChildOf(this.backgroundBlock);

        backgroundBlock.setColor(color.toJavaColor());

        return this;
    }

    public PSSButton setDefaultColor() {
        backgroundBlock.removeChild(buttonTexture);

        buttonTexture = (UIImage) ThemeManager.getCurrentButtonUIImage()
            .setWidth(new PixelConstraint(this.width))
            .setHeight(new PixelConstraint(this.height))
            .setX(new CenterConstraint())
            .setY(new CenterConstraint())
            .setChildOf(this.backgroundBlock);

        backgroundBlock.setColor(new Color(0, 0, 0, 0));

        this.color = new OneColor(0, 0, 0, 0);
        return this;
    }

    public PSSButton setText(String text) {
        this.textComponent.setText(text);
        this.text = text;

        return this;
    }

    public PSSButton setTextScale(float scale) {
        textComponent.setTextScale(new PixelConstraint(scale));

        return this;
    }

    public PSSButton onMouseClickConsumer(Consumer<UIClickEvent> method) {
        backgroundBlock.onMouseClickConsumer(method);

        return this;
    }


    public UIComponent getComponent() {
        return buttonTexture;
    }

    public PSSButton setBackgroundVisibility(boolean val) {
        if (val) {
            buttonTexture.unhide(true);
            for (UIComponent child : buttonTexture.getChildren()) {
                child.unhide(true);
            }
        } else {
            buttonTexture.hide();
            for (UIComponent child : buttonTexture.getChildren()) {
                child.unhide(true);
            }
        }

        return this;
    }

    public PSSButton insertComponentBeforeBackground(UIComponent component) {
        backgroundBlock.insertChildBefore(component, buttonTexture);

        return this;
    }
}

File: src/main/kotlin/me/partlysanestudios/partlysaneskies/render/gui/components/PSSButton.kt
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//
package me.partlysanestudios.partlysaneskies.render.gui.components

import cc.polyfrost.oneconfig.config.core.OneColor
import gg.essential.elementa.UIComponent
import gg.essential.elementa.components.UIBlock
import gg.essential.elementa.components.UIWrappedText
import gg.essential.elementa.constraints.CenterConstraint
import gg.essential.elementa.constraints.HeightConstraint
import gg.essential.elementa.constraints.WidthConstraint
import gg.essential.elementa.constraints.XConstraint
import gg.essential.elementa.constraints.YConstraint
import gg.essential.elementa.dsl.childOf
import gg.essential.elementa.dsl.constrain
import gg.essential.elementa.dsl.constraint
import gg.essential.elementa.dsl.percent
import gg.essential.elementa.events.UIClickEvent
import me.partlysanestudios.partlysaneskies.features.themes.ThemeManager.currentButtonUIImage
import me.partlysanestudios.partlysaneskies.features.themes.ThemeManager.getCurrentButtonUIImage
import java.awt.Color
import java.util.function.Consumer

class PSSButton {

    private var color: OneColor
    private var text: String = ""
    private val onMouseClick = ArrayList<Consumer<UIClickEvent>>()

    private val backgroundBlock = UIBlock().constrain {
        color = Color(0, 0, 0, 0).constraint
    }.onMouseClick {
        for (method in onMouseClick) {
            method.accept(it)
        }
    }

    private var buttonTexture = currentButtonUIImage.constrain {
        x = CenterConstraint()
        y = CenterConstraint()
        width = 100.percent
        height = 100.percent
    }.onMouseEnter {
        this.setWidth(105.percent)
        this.setHeight(105.percent)
    }.onMouseLeave {
        this.setWidth(100.percent)
        this.setHeight(100.percent)
    } childOf backgroundBlock

    private val textComponent = UIWrappedText(text, false, Color(0, 0, 0, 0), true).constrain {
        x = CenterConstraint()
        y = CenterConstraint()
        width = 100.percent
        color = Color.lightGray.constraint
    } childOf backgroundBlock

    constructor() {
        text = ""
        color = OneColor(0, 0, 0, 0)
    }

    constructor(color: Color) : this(OneColor(color))

    constructor(color: OneColor) {
        text = ""
        backgroundBlock.setColor(color.toJavaColor())
        this.color = color
    }

    fun setHeight(height: HeightConstraint): PSSButton {
        backgroundBlock.setHeight(height)
        return this
    }

    fun setWidth(width: WidthConstraint): PSSButton {
        backgroundBlock.setWidth(width)
        return this
    }

    fun setX(xPos: XConstraint): PSSButton {
        backgroundBlock.setX(xPos)
        return this
    }

    fun setY(yPos: YConstraint): PSSButton {
        backgroundBlock.setY(yPos)
        return this
    }

    fun setChildOf(parent: UIComponent): PSSButton {
        backgroundBlock.setChildOf(parent)
        return this
    }

    fun setColor(color: Color): PSSButton {
        return setColor(OneColor(color))
    }

    fun setColor(color: OneColor): PSSButton {
        backgroundBlock.removeChild(buttonTexture)
        buttonTexture = getCurrentButtonUIImage(color).constrain {
            x = CenterConstraint()
            y = CenterConstraint()
            width = 100.percent
            height = 100.percent
        } childOf backgroundBlock

        textComponent childOf buttonTexture

        backgroundBlock.setColor(color.toJavaColor())
        return this
    }

    fun setDefaultColor(): PSSButton {
        backgroundBlock.removeChild(buttonTexture)
        buttonTexture = currentButtonUIImage.constrain {
            x = CenterConstraint()
            y = CenterConstraint()
            width = 100.percent
            height = 100.percent
        } childOf backgroundBlock

        textComponent childOf buttonTexture

        backgroundBlock.setColor(Color(0, 0, 0, 0))
        return this
    }

    fun setText(text: String): PSSButton {
        textComponent.setText(text)
        this.text = text
        return this
    }

    fun setTextScale(scale: HeightConstraint): PSSButton {
        textComponent.setTextScale(scale)
        return this
    }

    fun onMouseClickConsumer(method: Consumer<UIClickEvent>): PSSButton {
        onMouseClick.add(method)
        return this
    }

    val component: UIComponent
        get() = buttonTexture

    fun setBackgroundVisibility(value: Boolean): PSSButton {
        if (value) {
            buttonTexture.unhide(true)
            for (child in buttonTexture.children) {
                child.unhide(true)
            }
        } else {
            buttonTexture.hide()
            for (child in buttonTexture.children) {
                child.unhide(true)
            }
        }
        return this
    }

    fun insertComponentBeforeBackground(component: UIComponent): PSSButton {
        backgroundBlock.insertChildBefore(component, buttonTexture)
        return this
    }
}
